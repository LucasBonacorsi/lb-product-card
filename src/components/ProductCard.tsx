import styles from "../styles/styles.module.css";
import { useProduct } from "../hooks/useProduct";
import React,{ createContext, CSSProperties, } from "react";
import {
  InitialValues,
  OnChangeArgs,
  Product,
  ProductCardHandlers,
  ProductContextProps,
} from "../interfaces/interfaces";
import { ProductImage } from "./ProductImage";
import { ProductTitle } from "./ProductTitle";
import ProductButtons from "./ProductButtons";

export const ProductContext = createContext({} as ProductContextProps);

export interface ProductCardProps {
  product: Product;
  children: ((args: ProductCardHandlers) => JSX.Element);
  className?: string;
  style?: CSSProperties;
  onChange?: (args: OnChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductCard = ({
  product,
  children,
  className,
  style,
  onChange,
  value,
  initialValues,
}: ProductCardProps) => {
  const { counter, increaseBy,
    isMaxCountReached,reset} = useProduct({
    onChange,
    product,
    value,
    initialValues
  });

  return (
    <ProductContext.Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount: initialValues?.maxCount
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          isMaxCountReached,
          reset,
          maxCount: initialValues?.maxCount,
          product,
          increaseBy
        })}
      </div>
    </ProductContext.Provider>
  );
};

ProductCard.Title = ProductTitle;
ProductCard.Image = ProductImage;
ProductCard.Button = ProductButtons;
