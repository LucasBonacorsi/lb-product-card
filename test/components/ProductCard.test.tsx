import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { ProductCard } from '../../src/components/index';
import { product1 } from '../data/products';

describe('ProductCard', () => {
  test('Debe mostrar el componente correctament con el titulo ', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {() => <h1>Product card</h1>}
      </ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('Debe mostrar el componente correctament con el titulo ', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {({ count, increaseBy }) => (
          <>
            <h1>Product card</h1>
            <span>{count}</span>
            <button onClick={() => increaseBy(1)}></button>
          </>
        )}
      </ProductCard>
    );

    let tree = wrapper.toJSON()
    console.log(tree);
    expect(tree).toMatchSnapshot();

    act(() => { 
      (tree as any).children[2].props.onClick()
     })
     tree = wrapper.toJSON();
    expect((tree as any).children[1].children[0]).toBe('1');
  });
});
