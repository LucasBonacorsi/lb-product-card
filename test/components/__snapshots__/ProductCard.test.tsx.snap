// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard Debe mostrar el componente correctament con el titulo  1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product card
  </h1>
</div>
`;

exports[`ProductCard Debe mostrar el componente correctament con el titulo  2`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  />
</div>
`;
